trieste/models/gpflow/sampler.py

```diff
@@ -190,32 +196,133 @@ def sample(self, at: TensorType, *, jitter: float = DEFAULTS.JITTER) -> TensorTy
+        """
+
+    def get_trajectory(self) -> TrajectoryFunction:
+        """
+        Generate an approximate function draw (trajectory) by sampling weights
+        and evaluating the feature functions.
+
+        :return: A trajectory function representing an approximate trajectory from the Gaussian
+            process, taking an input of shape `[N, D]` and returning shape `[N, 1]`
+        """
+
+        weight_sampler = self._prepare_weight_sampler()  # prep feature weight distribution
+
+        return feature_decomposition_trajectory(
+            feature_functions=self._feature_functions,
+            weight_sampler=weight_sampler,
+        )
+
+    def update_trajectory(self, trajectory: TrajectoryFunction) -> TrajectoryFunction:
+        """
+        Efficiently update a :const:`TrajectoryFunction` to reflect an update in its
+        underlying :class:`ProbabilisticModel` and resample accordingly.
+
+        For a :class:`FeatureDecompositionTrajectorySampler`, updating the sampler
+        corresponds to resampling the feature functions (taking into account any
+        changed kernel parameters) and recalculating the weight distribution.
+
+        :param trajectory: The trajectory function to be resampled.
+        :return: The new resampled trajectory function.
+        """
+        tf.debugging.Assert(isinstance(trajectory, feature_decomposition_trajectory), [])
+
+        self._feature_functions.resample()  # resample Fourier feature decomposition
+        weight_sampler = self._prepare_weight_sampler()  # recalculate weight distribution
+
+        trajectory.update(weight_sampler=weight_sampler)  # type: ignore
+
+        return trajectory  # return trajectory with updated features and weight distribution
+
+    def resample_trajectory(self, trajectory: TrajectoryFunction) -> TrajectoryFunction:
+        """
+        Efficiently resample a :const:`TrajectoryFunction` in-place to avoid function retracing
+        with every new sample.
+
+        :param trajectory: The trajectory function to be resampled.
+        :return: The new resampled trajectory function.
+        """
+        tf.debugging.Assert(isinstance(trajectory, feature_decomposition_trajectory), [])
+        trajectory.resample()  # type: ignore
+        return trajectory  # return trajectory with resampled weights
+
+    @abstractmethod
+    def _prepare_weight_sampler(self) -> Callable[[int], TensorType]:  # [B] -> [L, B]
+        """
+        Calculate the posterior of the feature weights for the specified feature functions,
+        returning a function that takes in a batch size `B` and returns `B` samples for
+        the weights of each of the `L` features.
+        """
+        raise NotImplementedError
+
+
 @runtime_checkable
-class SupportsGetKernelObservationNoiseInternalData(
-    SupportsGetKernel, SupportsGetObservationNoise, SupportsInternalData, Protocol
+class SupportsGetKernelGetObservationNoiseSupportsGetInternalData(
```

Comment:
If you think the class name is too unwieldy you can rename it to something like:
```suggestion
class RandomFourierFeatureTrajectorySamplerModel(
```